### PR TITLE
[V9]Stock Picking Invoicing

### PR DESCRIPTION
**Pull request by flotho**
Since Odoo v9 doesn't support anymore invoicing neither on move nor on picking, this migration has the purpose to summarize some modules :
- https://github.com/OCA/account-invoicing/tree/9.0/stock_picking_invoicing
- https://github.com/OCA/stock-logistics-workflow/tree/9.0/stock_picking_invoice_link
- https://github.com/OCA/account-invoicing/tree/8.0/stock_picking_invoicing_unified
- And the odoo core one : https://github.com/odoo/odoo/blob/8.0/addons/stock_account/stock.py

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I put the label "Work in progress" until you notify the contrary.

**Comment by flotho** [comment]
Here we are,

@pedrobaeza could yo put the need review tag ?

thanks

File: stock_picking_invoicing/README.rst [hunk 1]
```diff
@@ -25,6 +25,7 @@ Contributors
 * Leonardo Pistone <leonardo.pistone@agilebg.com>
 * Daniel Sadamo <daniel.sadamo@kmee.com.br>
 * Alex Comba <alex.comba@agilebg.com>
+* Florent THOMAS <florent.thomas@mind-and-go.com>
```
Comment (elicoidal) [inline stock_picking_invoicing/README.rst:71]: @flotho l  Thanks for your contribution to the OCA. 
Please follow the principles: 
- Contribution guidelines: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
- Can you please update to the newest OCA README template:
  https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
- We recommend to include a screenshot as well to help new comers (nice to have)
- Please use simple headers in all py files following OCA templates
  https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

File: stock_picking_invoicing/__init__.py [hunk 2]
```diff
@@ -18,4 +18,5 @@
 #
 ##############################################################################
 
-from . import stock
```
Comment (elicoidal) [inline stock_picking_invoicing/__init__.py:21]: please update all copyright headers to the simplified format

File: stock_picking_invoicing/models/stock.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+import logging
+from openerp import _
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py]: one line with all imports?

File: stock_picking_invoicing/models/stock.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+
+
+class StockLocationPath(models.Model):
+    _inherit = "stock.location.path"
+    
+    invoice_state = fields.Selection(INVOICE_STATE, 
+                                     string="Invoice Status", default="none")
+    
+    
+    @api.v7
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:25]: I understand this is a forward port from openerp code but shouldnot we take the opportunity to migrate to new API?

File: stock_picking_invoicing/models/stock.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+
+#----------------------------------------------------------
+# Move
+#----------------------------------------------------------
+
+class stock_move(models.Model):
+    _inherit = "stock.move"
+
+    #=====================
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:56]: those comments are not necessary

File: stock_picking_invoicing/models/stock.py (hunk 4 above)
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:25]: sorry: probably already the case!

File: stock_picking_invoicing/models/stock.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+            code = self.get_code_from_locs(move)
+            if partner.property_product_pricelist and code == 'outgoing':
+                currency = partner.property_product_pricelist.currency_id.id
+        data = partner, self.env.uid, currency
+        
+        if move.picking_id.partner_id.id != data[0].id:
+            # if someone else modified invoice partner, I use it
+            return data
+        
+        partner_invoice_id = move.picking_id.partner_id.address_get(
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:93]: improve indentation

File: stock_picking_invoicing/models/stock.py [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+            if not account_id:
+                account_id = self.product_id.categ_id.property_account_expense_categ_id.id
+        fiscal_position = fp or partner.property_account_position_id
+        
+        account_id = self._get_account(fiscal_position, account_id)  
+
+        # set UoS if it's a sale and the picking doesn't have one
+        uos_id = self.product_uom.id
+        quantity = self.product_uom_qty
+#        if move.product_uos:
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:142]: remove dead code

File: stock_picking_invoicing/models/stock.py [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+            'uos_id': uos_id,
+            'uom_id': uos_id,
+            'quantity': quantity,
+            'price_unit': self._get_price_unit_invoice(inv_type),
+            'invoice_line_tax_ids': [(6, 0, taxes_ids)],
+            'discount': 0.0,
+            'account_analytic_id': False,
+        }
+
+        #https://github.com/OCA/account-invoicing/blob/8.0/stock_picking_invoicing_unified/models/stock_move.py
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py]: Please explain in your comment the link

File: stock_picking_invoicing/models/stock.py [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+        @return: The price unit for the move line
+        """
+        
+        if type in ('in_invoice', 'in_refund'):
+            return move_line.product_id.price        
+        else:
+            # If partner given, search price in its sale pricelist
+            if self.partner_id and self.partner_id.property_product_pricelist:
+                self = self.with_context(
+                                         partner=self.partner_id.id,
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:191]: improve (reduce) identation

File: stock_picking_invoicing/models/stock.py [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+
+    #===========================
+    # Views
+    #===========================
+    
+    @api.multi
+    def set_to_be_invoiced(self):
+        for picking in self:
+            if picking.invoice_state == '2binvoiced':
+                raise Warning(_("Can't update invoice control for picking %s: "
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py]: Invoice control cannot be updated for picking %s as it is already set as "To be invoiced"

File: stock_picking_invoicing/models/stock.py [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+                              "It's 'to be invoiced' yet") % picking.name)
+            if picking.invoice_state in ('none', 'invoiced'):
+                if picking.invoice_id:
+                    raise Warning(_('Picking %s has linked invoice %s') %
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py]: Picking %s is already linked to invoice %s

File: stock_picking_invoicing/models/stock.py [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+            for move in picking.move_lines:
+                if move.invoice_state != 'invoiced':
+                    move.invoice_state = '2binvoiced'
+        return True
+    
+    @api.multi
+    def set_invoiced(self):
+        for picking in self:
+            if picking.invoice_state == 'invoiced' or picking.invoice_id :
+                raise Warning(_("Can't update invoice control for picking %s: "
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py]: Invoice control cannot be updated for picking %s as it is already set as "Invoiced"

File: stock_picking_invoicing/models/stock.py [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+                              "It's already invoiced!") % picking.name)
+            if picking.invoice_state in ('2binvoiced'):
+                if picking.invoice_id:
+                    raise Warning(_('Picking %s has linked invoice %s') %
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py]: Picking %s is already linked to invoice %s

File: stock_picking_invoicing/models/stock.py [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+                if move.invoice_state != 'invoiced':
+                    move.invoice_state = '2binvoiced'
+        return True
+
+    #=====================
+    # FIELDS
+    #=====================
+
+    invoice_state = fields.Selection(INVOICE_STATE, 
+                                     string="Invoice Status"
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:261]: reduce indentation

File: stock_picking_invoicing/models/stock.py [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+                    merge_vals['origin'] = ', '.join(invoice_origin)
+                    
+                if invoice_vals.get('name', False) and (not invoice.name or invoice_vals['name'] not in invoice.name.split(', ')):
+                    invoice_name = filter(None, [invoice.name, invoice_vals['name']])
+                    merge_vals['name'] = ', '.join(invoice_name)
+                if merge_vals:
+                    invoice.write(merge_vals)
+
+            move.with_context(
+                              fp_id=invoice_vals.get('fiscal_position_id', False)
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py]: reduce indentation

File: stock_picking_invoicing/models/stock.py [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+                })
+            if move.picking_id and not move.picking_id.invoice_id :
+                move.picking_id.write({'invoice_id': invoice_id.id,
+                                    'invoice_state': 'invoiced'
+                })
+                
+        
+        invoice_id.compute_invoice_tax_lines()
+
+#        invoice_obj.button_compute(cr, uid, invoices.values(), context=context, set_total=(inv_type in ('in_invoice', 'in_refund')))
```
Comment (elicoidal) [inline stock_picking_invoicing/models/stock.py:389]: remove

File: stock_picking_invoicing/views/stock_view.xml [hunk 17]
```diff
@@ -0,0 +1,132 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+
```
Comment (elicoidal) [inline stock_picking_invoicing/views/stock_view.xml]: remove empty lines

File: stock_picking_invoicing/wizard/stock_invoice_onshipping.py [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+        
+        res = super(StockInvoiceOnshipping, self).view_init(fields)
+        pick_obj = self.env['stock.picking']
+        count = 0
+        active_ids = context.get('active_ids',[])        
+        for pick in pick_obj.search([('id', 'in', active_ids)]):
+            if pick.invoice_state != '2binvoiced':
+                count += 1
+            if not pick.partner_id :
+                raise Warning(_('All your picking must have a partner to be invoiced!'))
```
Comment (elicoidal) [inline stock_picking_invoicing/wizard/stock_invoice_onshipping.py]: s/picking/pickings

File: stock_picking_invoicing/wizard/stock_invoice_onshipping.py [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+        if len(active_ids) == count:
+            _logger.debug("Raise ")
+            raise Warning(_('None of these picking lists require invoicing.'))
```
Comment (elicoidal) [inline stock_picking_invoicing/wizard/stock_invoice_onshipping.py]: None of these picking requires invoicing.

File: stock_picking_invoicing/wizard/stock_invoice_onshipping.py [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+    def onchange_group(self):
+        self.ensure_one()
+        (sale_pickings, sale_refund_pickings, purchase_pickings,
+            purchase_refund_pickings) = self.get_split_pickings()
+        self.show_sale_journal = bool(sale_pickings)
+        self.show_sale_refund_journal = bool(sale_refund_pickings)
+        self.show_purchase_journal = bool(purchase_pickings)
+        self.show_purchase_refund_journal = bool(purchase_refund_pickings)
+
+#    @api.onchange('journal_id')
```
Comment (elicoidal) [inline stock_picking_invoicing/wizard/stock_invoice_onshipping.py]: remove dead code

File: stock_picking_invoicing/wizard/stock_invoice_onshipping.py [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+        usage = pick.move_lines[0].location_id.usage if type == 'incoming' else pick.move_lines[0].location_dest_id.usage
+
+        return JOURNAL_TYPE_MAP.get((type, usage), ['sale'])[0]
+    
+    #=======================
+    # FIELDS
+    #=======================
+
+    journal_id = fields.Many2one(comodel_name='account.journal'
+                                , string= 'Destination Journal'
```
Comment (elicoidal) [inline stock_picking_invoicing/wizard/stock_invoice_onshipping.py:129]: indentation

File: stock_picking_invoicing/models/stock.py (hunk 6 above)
Comment (flotho) [inline stock_picking_invoicing/models/stock.py:93]: What are the recommendations?

File: stock_picking_invoicing/models/stock.py (hunk 7 above)
Comment (flotho) [inline stock_picking_invoicing/models/stock.py:142]: You mean the one commented?

File: stock_picking_invoicing/models/stock.py (hunk 8 above)
Comment (flotho) [inline stock_picking_invoicing/models/stock.py]: Removed, it was a reminder of the source code

File: stock_picking_invoicing/models/stock.py (hunk 14 above)
Comment (flotho) [inline stock_picking_invoicing/models/stock.py:261]: What are the recommendations?

**Pull request closed without merging**